Allow specifying reporters on the CLI

mischnic opened the pull request:
```sh
yarn parcel build src/index.js --reporter @parcel/reporter-bundle-buddy
```

- [x] Is `options.reporters` is misleading? Maybe `additionalReporters` is better?

Closes T-786

parcel-benchmark commented:
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.30s | -89.00ms |
| Cached | 463.00ms | +5.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 96.00ms | +24.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 124.00ms | +21.00ms ⚠️ |
| dist/legacy/index.html | 701.00b | +0.00b | 645.00ms | +64.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 69.00ms | -6.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 50.00ms | -3.00ms 🚀 |
| dist/legacy/index.76ce4b0c.js | 1.12kb | +0.00b | 70.00ms | -5.00ms 🚀 |
| dist/modern/index.b43fcc94.js | 1.12kb | +0.00b | 70.00ms | -5.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 73.00ms | -5.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 70.00ms | -5.00ms 🚀 |
| dist/legacy/index.44edca46.css | 77.00b | +0.00b | 70.00ms | -5.00ms 🚀 |
| dist/modern/index.0bc656bc.css | 77.00b | +0.00b | 70.00ms | -5.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 20.38s | +471.00ms |
| Cached | 950.00ms | -7.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/NotFound.f3269ae7.js | 542.00b | +0.00b | 312.00ms | +69.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 71.00ms | -10.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/NotFound.f3269ae7.js | 542.00b | +0.00b | 84.00ms | -38.00ms 🚀 |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 84.00ms | -11.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.02m | +275.00ms |
| Cached | 3.15s | +51.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/popup.119ff481.js | 171.12kb | +0.00b | 15.95s | -849.00ms 🚀 |
| dist/EmojiPickerComponent.e20a59f3.js | 139.77kb | +0.00b | 10.82s | -1.53s 🚀 |
| dist/media-viewer.ebcb5e1e.js | 73.43kb | +0.00b | 11.30s | -1.42s 🚀 |
| dist/card.1d931b4c.js | 52.43kb | +0.00b | 12.42s | -1.57s 🚀 |
| dist/Modal.558fb5db.js | 39.59kb | +0.00b | 3.79s | -470.00ms 🚀 |
| dist/component.ad27194c.js | 31.05kb | +0.00b | 4.02s | +349.00ms ⚠️ |
| dist/esm.e3419f0a.js | 27.83kb | +0.00b | 9.42s | -2.92s 🚀 |
| dist/component.f25dcadb.js | 22.80kb | +0.00b | 3.77s | -473.00ms 🚀 |
| dist/DatePicker.faca45d0.js | 21.38kb | +0.00b | 9.40s | -768.00ms 🚀 |
| dist/smartMediaEditor.21f91ed8.js | 16.96kb | +0.00b | 16.33s | +2.14s ⚠️ |
| dist/js.8167e2d5.js | 16.55kb | +0.00b | 3.77s | -473.00ms 🚀 |
| dist/workerHasher.bc6346e1.js | 11.90kb | +0.00b | 7.27s | -500.00ms 🚀 |
| dist/component.677fad45.js | 6.23kb | +0.00b | 4.29s | +343.00ms ⚠️ |
| dist/media-viewer.f15e56f4.js | 4.01kb | +0.00b | 11.34s | -2.26s 🚀 |
| dist/EmojiPickerComponent.4991ddd4.js | 3.68kb | +0.00b | 7.15s | -1.29s 🚀 |
| dist/png-chunks-extract.608f438d.js | 3.56kb | +0.00b | 3.77s | -473.00ms 🚀 |
| dist/dropzone.d1e8004d.js | 3.46kb | +0.00b | 10.99s | -1.35s 🚀 |
| dist/Modal.08282ab6.js | 3.23kb | +0.00b | 1.09s | -62.00ms 🚀 |
| dist/clipboard.6a6e5fbd.js | 3.00kb | +0.00b | 15.66s | +2.04s ⚠️ |
| dist/16.b6ca7fd0.js | 2.50kb | +0.00b | 1.09s | -63.00ms 🚀 |
| dist/ResourcedEmojiComponent.5046c0ff.js | 2.19kb | +0.00b | 8.06s | -1.83s 🚀 |
| dist/date.a69000aa.js | 1.97kb | +0.00b | 4.38s | -387.00ms 🚀 |
| dist/feedback.facaac2e.js | 1.87kb | +0.00b | 9.38s | -642.00ms 🚀 |
| dist/16.a9ea9b41.js | 1.86kb | +0.00b | 4.11s | -356.00ms 🚀 |
| dist/browser.8883c452.js | 1.85kb | +0.00b | 15.81s | -834.00ms 🚀 |
| dist/16.89d45d15.js | 1.80kb | +0.00b | 3.94s | -284.00ms 🚀 |
| dist/workerHasher.31326e89.js | 1.77kb | +0.00b | 5.40s | -2.34s 🚀 |
| dist/workerHasher.d607e0b8.js | 1.77kb | +0.00b | 12.43s | -4.40s 🚀 |
| dist/list-number.d4568b49.js | 1.69kb | +0.00b | 4.48s | -386.00ms 🚀 |
| dist/code.efde0f48.js | 1.61kb | +0.00b | 4.29s | -401.00ms 🚀 |
| dist/link.5be155be.js | 1.54kb | +0.00b | 4.48s | -387.00ms 🚀 |
| dist/heading6.5a6082c8.js | 1.53kb | +0.00b | 4.72s | -2.61s 🚀 |
| dist/heading3.1c92891c.js | 1.52kb | +0.00b | 4.72s | -377.00ms 🚀 |
| dist/16.07ccad90.js | 1.51kb | +0.00b | 4.11s | -671.00ms 🚀 |
| dist/16.00f25203.js | 1.47kb | +0.00b | 3.94s | -284.00ms 🚀 |
| dist/16.02056ae6.js | 1.47kb | +0.00b | 3.95s | -830.00ms 🚀 |
| dist/emoji.b5624cf5.js | 1.46kb | +0.00b | 4.38s | -389.00ms 🚀 |
| dist/16.b7da7bde.js | 1.45kb | +0.00b | 1.09s | -62.00ms 🚀 |
| dist/16.00100306.js | 1.44kb | +0.00b | 4.26s | -382.00ms 🚀 |
| dist/16.bdd61d2d.js | 1.41kb | +0.00b | 4.26s | -815.00ms 🚀 |
| dist/heading5.88bc6249.js | 1.41kb | +0.00b | 4.72s | -509.00ms 🚀 |
| dist/expand.4a483eb0.js | 1.39kb | +0.00b | 9.38s | -688.00ms 🚀 |
| dist/16.2a21d3f4.js | 1.34kb | +0.00b | 1.09s | -62.00ms 🚀 |
| dist/heading2.22de70fe.js | 1.34kb | +0.00b | 4.72s | -377.00ms 🚀 |
| dist/16.c9db7e6b.js | 1.32kb | +0.00b | 4.26s | -382.00ms 🚀 |
| dist/mention.c44b48ed.js | 1.30kb | +0.00b | 4.48s | -1.74s 🚀 |
| dist/heading4.a2524c70.js | 1.29kb | +0.00b | 4.72s | -377.00ms 🚀 |
| dist/Modal.44e980e0.js | 1.29kb | +0.00b | 3.77s | -473.00ms 🚀 |
| dist/layout.81f2196c.js | 1.28kb | +0.00b | 4.38s | -482.00ms 🚀 |
| dist/16.9e97e787.js | 1.28kb | +0.00b | 1.09s | -63.00ms 🚀 |
| dist/16.acff901d.js | 1.27kb | +0.00b | 4.11s | -360.00ms 🚀 |
| dist/component.233c7a65.js | 1.26kb | +0.00b | 4.02s | +349.00ms ⚠️ |
| dist/16.015dfb9d.js | 1.26kb | +0.00b | 3.94s | -525.00ms 🚀 |
| dist/16.828b958e.js | 1.26kb | +0.00b | 4.11s | -671.00ms 🚀 |
| dist/16.b9b1643b.js | 1.26kb | +0.00b | 4.11s | -357.00ms 🚀 |
| dist/16.ec84a250.js | 1.26kb | +0.00b | 3.94s | -284.00ms 🚀 |
| dist/divider.ed7ccd2e.js | 1.26kb | +0.00b | 4.38s | -389.00ms 🚀 |
| dist/quote.b6d8bffb.js | 1.25kb | +0.00b | 4.64s | -338.00ms 🚀 |
| dist/media-card-analytics-error-boundary.ff0177b3.js | 1.25kb | +0.00b | 10.09s | -818.00ms 🚀 |
| dist/action.01ec8983.js | 1.24kb | +0.00b | 4.26s | -382.00ms 🚀 |
| dist/16.458399e0.js | 1.23kb | +0.00b | 4.02s | +349.00ms ⚠️ |
| dist/decision.c48f9f1e.js | 1.22kb | +0.00b | 4.38s | -389.00ms 🚀 |
| dist/panel-warning.b5f3360d.js | 1.22kb | +0.00b | 4.64s | -337.00ms 🚀 |
| dist/list.2dfe3f18.js | 1.19kb | +0.00b | 4.48s | -386.00ms 🚀 |
| dist/panel-error.7454ebc0.js | 1.12kb | +0.00b | 4.49s | -1.74s 🚀 |
| dist/panel.fc70749c.js | 1.11kb | +0.00b | 4.64s | -341.00ms 🚀 |
| dist/panel-success.d5752da0.js | 1.06kb | +0.00b | 4.64s | -249.00ms 🚀 |
| dist/panel-note.6a71a1f2.js | 1.05kb | +0.00b | 4.64s | -1.58s 🚀 |
| dist/media-picker-analytics-error-boundary.49b32f4d.js | 1023.00b | +0.00b | 15.63s | +3.04s ⚠️ |
| dist/media-card-analytics-error-boundary.b884d580.js | 1019.00b | +0.00b | 12.42s | -4.31s 🚀 |
| dist/simpleHasher.79137cb4.js | 767.00b | +0.00b | 7.15s | -591.00ms 🚀 |
| dist/index.html | 119.00b | +0.00b | 2.92s | +200.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.538ccbed.js | 2.46mb | +0.00b | 597.00ms | -33.00ms 🚀 |
| dist/component.ad27194c.js | 31.05kb | +0.00b | 598.00ms | +40.00ms ⚠️ |
| dist/component.f25dcadb.js | 22.80kb | +0.00b | 606.00ms | +36.00ms ⚠️ |
| dist/js.8167e2d5.js | 16.55kb | +0.00b | 606.00ms | +36.00ms ⚠️ |
| dist/png-chunks-extract.608f438d.js | 3.56kb | +0.00b | 606.00ms | +36.00ms ⚠️ |
| dist/Modal.08282ab6.js | 3.23kb | +0.00b | 597.00ms | -33.00ms 🚀 |
| dist/16.b6ca7fd0.js | 2.50kb | +0.00b | 597.00ms | +39.00ms ⚠️ |
| dist/16.b7da7bde.js | 1.45kb | +0.00b | 597.00ms | +39.00ms ⚠️ |
| dist/16.2a21d3f4.js | 1.34kb | +0.00b | 597.00ms | +39.00ms ⚠️ |
| dist/Modal.44e980e0.js | 1.29kb | +0.00b | 606.00ms | +37.00ms ⚠️ |
| dist/16.9e97e787.js | 1.28kb | +0.00b | 597.00ms | +39.00ms ⚠️ |
| dist/component.233c7a65.js | 1.26kb | +0.00b | 597.00ms | -32.00ms 🚀 |
| dist/16.458399e0.js | 1.23kb | +0.00b | 598.00ms | -31.00ms 🚀 |
| dist/index.html | 119.00b | +0.00b | 523.00ms | -48.00ms 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 14.71s | -179.00ms |
| Cached | 542.00ms | -5.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/291222443189600777)**

DeMoorJasper commented on packages/core/core/src/resolveOptions.js:
```diff
@@ -120,6 +120,7 @@ export default async function resolveOptions(
     outputFS,
     cache,
     packageManager,
+    reporters: initialOptions.reporters ?? [],
```
Would personally prefer this to be more explicit like additionalReporters, not sure what other ppl think about this

devongovett replied:
Should reporters be in .parcelrc at all or should they just be an API/CLI thing?

mischnic replied:
That would make sense for all of the current official reporters, they are all either CLI related or only used in development.
But there are also cases like https://www.npmjs.com/package/@kicker-matchday/parcel-reporter-bundle-manifest (where you'd still want to run them every time). 

So one possibility would still be allowing reporters in the config but we could pass the cli and dev-server reporter from the CLI.

devongovett replied:
Yeah I like that. Then we could also only add the dev server reporter in parcel serve.

mischnic replied:
The problem with that is though: where are the reporters resolved from? We'd have to always autoinstall them

devongovett replied:
Yeah. How do we do it for the defaultConfig option. I guess it should match?

mischnic replied:
`defaultConfig` is just a filepath to the JSON file. Where would we get the reporter package name from? Reading the name in package.json relative to the filepath?

joeyslater commented on packages/core/parcel/src/cli.js:
```diff
@@ -418,6 +430,30 @@ async function normalizeOptions(command): Promise<InitialParcelOptions> {
+        potentialNames.join(', '),
+    );
+  }
+  console.log(reporters);
```
```suggestion
  INTERNAL_ORIGINAL_CONSOLE.log(reporters);
```

if you were intending to report the logs

mischnic replied:
Looks like a leftover from debugging 🙈

wbinnssmith commented:
Is it possible to override reporters entirely? For example, Jest has a shorthand for referencing the default set of reporters (`--reporters "default"`) that would be used: https://jestjs.io/docs/en/cli#--reporters

Use case: The JSON reporter would want stdio to itself, but currently the default config uses the default cli reporter.

mischnic commented:
> Use case: The JSON reporter would want stdio to itself, but currently the default config uses the default cli reporter.

We could pass cli reporter from the `parcel` CLI as I've said here: https://github.com/parcel-bundler/parcel/pull/5905#discussion_r580265493.

wbinnssmith commented:
> We could pass cli reporter from the parcel CLI as I've said here: #5905 (comment).

This would still make it difficult to use e.g. the json reporter from the cli for piping to other tools without re-implementing the Parcel cli, right?

mischnic commented:
Right, I was thinking from the `new Parcel` API perspective, not the CLI.

Passing the two reporters from the CLI would indeed make it impossible to turn them off via
```json
{
  "extends": "...",
  "reporters": []
}
```
so that's one downside of that.

mischnic commented:
I've moved added `additionalReporters.resolveFrom` and moved the two default reporters into the CLI.

Will do a followup for `parcel --json` which swaps out reporter-cli with reporter-json

devongovett commented on packages/core/parcel/src/cli.js:
```diff
@@ -418,6 +430,35 @@ async function normalizeOptions(command): Promise<InitialParcelOptions> {
+      name,
+      `@parcel/reporter-${name}`,
+      `parcel-reporter-${name}`,
+    ];
```
Hmmmm what do we think about this? We don't support these shorthands anywhere else...

mischnic replied:
(I did this because `parcel build src/index.html --reporter @parcel/reporter-bundle-buddy` can lead to either very long commands and it's just faster to type just the last part for a one-off build. But I don't feel strongly about it)

devongovett replied:
I'm ok with it just pointing out the difference. Maybe post in slack and get other opinions?

devongovett commented on packages/core/parcel/src/cli.js:
```diff
@@ -423,6 +435,15 @@ async function normalizeOptions(command): Promise<InitialParcelOptions> {
 
+  let additionalReporters = [
+    {packageName: '@parcel/reporter-cli', resolveFrom: __filename},
+    {packageName: '@parcel/reporter-dev-server', resolveFrom: __filename},
```
Technically we could only add this for serve mode but it's probably fine either way

devongovett commented on packages/core/core/src/requests/ParcelConfigRequest.js:
```diff
@@ -172,6 +172,17 @@ export async function resolveParcelConfig(
     options,
   );
+
+  if (options.additionalReporters.length > 0) {
```
Adding reporters shouldn't invalidate the cache. Can you make sure this option isn't tracked in the graph? I believe there is a list of ignored options.
